.github/workflows/test-go.yml

```diff
@@ -108,13 +113,29 @@ jobs:
       - name: List cached results
         id: list-cached-results
         run: ls -lhR test-results/go-test
-      - name: Build matrix excluding binary and integration tests
+      - name: Build matrix excluding binary, integration, and testonly tests
         id: build-non-binary
+        if: ${{ !inputs.testonly }}
         env:
           GOPRIVATE: github.com/hashicorp/*
         run: |
+          # testonly tests need additional build tag though let's exclude them anyway for clarity
+          (
+            go list ./... | grep -v "_binary" | grep -v "vault/integ" | grep -v "testonly" | gotestsum tool ci-matrix --debug \
+              --partitions "${{ inputs.total-runners }}" \
+              --timing-files 'test-results/go-test/*.json' > matrix.json
+          )
+      - name: Build matrix for tests tagged with testonly
+        if: ${{ inputs.testonly }}
+        env:
+          GOPRIVATE: github.com/hashicorp/*
+        run: |
+          set -exo pipefail
+          
+          # testonly tagged tests need an additional tag to be included
+          # also running some extra tests for sanity checking with the testonly build tag
           (
-            go list ./... | grep -v "_binary" | grep -v "vault/integ" | gotestsum tool ci-matrix --debug \
+            go list -tags=testonly ./... | grep -E -i 'external_tests/kv|external_tests/token|testonly|external_tests/replication-perf|^github.com/hashicorp/vault/vault$' | gotestsum tool ci-matrix --debug \
```

Comment:
We can probably save a few characters here by doing:
```
  ✭ ✱ ◼ [remotes/origin/testonly-ci-with-corruption-detection] ~/s/vault-enterprise >  go list -tags=testonly ./vault/external_tests/{kv,token,replication-perf,*testonly} ./vault/
github.com/hashicorp/vault/vault/external_tests/kv
github.com/hashicorp/vault/vault/external_tests/token
github.com/hashicorp/vault/vault/external_tests/replication-perf
github.com/hashicorp/vault/vault/external_tests/vault_testonly
github.com/hashicorp/vault/vault
```

Reply:
This does not work in OSS

Reply:
Of course 🐟 v. `sh`. We could use `shopt` but after discussing out-of-band this works as is, so let's use it 😄